src/tools/idlc/src/tt_create.c

```diff
@@ -548,6 +567,231 @@ extern void ddsts_struct_empty_close(ddsts_context_t *context, ddsts_type_t **re
+    case DDSTS_LONGDOUBLE:     return "long double"; break;
+    case DDSTS_FIXED_PT_CONST: return "fixed point const"; break;
+    case DDSTS_ANY:            return "any";
+    case DDSTS_SEQUENCE:       return "sequence";
+    case DDSTS_ARRAY:          return "array";
+    case DDSTS_STRING:         return "string";
+    case DDSTS_WIDE_STRING:    return "wide string";
+    case DDSTS_FIXED_PT:       return "fixed point";
+    case DDSTS_MAP:            return "map";
+    case DDSTS_MODULE:         return "module";
+    case DDSTS_FORWARD_STRUCT: return "forward struct";
+    case DDSTS_STRUCT:         return "struct";
+    case DDSTS_FORWARD_UNION:  return "forward union";
+    case DDSTS_UNION:          return "union";
+    default:
+      assert(0);
+      break;
+  }
+  return "";
+}
+
+static dds_return_t cast_value_to_type(ddsts_literal_t *value, ddsts_flags_t flags, ddsts_literal_t *cast_type)
+{
+  switch (value->flags) {
+    case DDSTS_ULONGLONG:
+      switch (flags) {
+        case DDSTS_OCTET:
+        case DDSTS_UINT8:
+        case DDSTS_USHORT:
+        case DDSTS_ULONG:
+        case DDSTS_ULONGLONG:
+          cast_type->value.ullng = value->value.ullng;
+          cast_type->flags = flags;
+          return DDS_RETCODE_OK;
+        case DDSTS_INT8:
+        case DDSTS_SHORT:
+        case DDSTS_LONG:
+          cast_type->value.llng = (signed long long)value->value.ullng;
+          cast_type->flags = flags;
+          return DDS_RETCODE_OK;
+        default:
+          DDS_ERROR("Cannot cast unsigned long long to %s\n", name_of_type(flags));
+          return DDS_RETCODE_ERROR;
+          break;
+      }
+      break;
+    default:
+      assert(0);
+      break;
+  }
+  return DDS_RETCODE_ERROR;
+}
+
+static bool equal_values(ddsts_literal_t *a, ddsts_literal_t *b)
+{
+  assert(a->flags == b->flags);
+  switch (a->flags) {
+    case DDSTS_OCTET:
+    case DDSTS_UINT8:
+    case DDSTS_USHORT:
+    case DDSTS_ULONG:
+    case DDSTS_ULONGLONG:
+      return a->value.ullng == b->value.ullng;
+      break;
+    case DDSTS_INT8:
+    case DDSTS_SHORT:
+    case DDSTS_LONG:
+      return a->value.llng == b->value.llng;
+      break;
+    default:
+      assert(0);
+      break;
+  }
+  return false;
+}
+
+extern bool ddsts_union_add_case_label(ddsts_context_t *context, ddsts_literal_t *value)
+{
+  assert(cur_scope_is_definition_type(context, DDSTS_UNION));
+  ddsts_literal_t casted_value;
+  dds_return_t rc = cast_value_to_type(value, context->cur_type->union_def.switch_type, &casted_value);
+  if (rc != DDS_RETCODE_OK) {
+    context->retcode = rc;
+    return false;
+  }
+  for (ddsts_type_t *cases = context->cur_type->union_def.cases.first; cases != NULL; cases = cases->type.next) {
+    for (ddsts_union_case_label_t *label = cases->union_case.labels; label != NULL; label = label->next) {
+      if (equal_values(&label->value, &casted_value)) {
+        DDS_ERROR("Same case label used in more than one case.\n");
```

Comment:
I don't see any mention of location in the source code, while `%locations` is enabled. I think that should be changed.

Reply:
The location is used for syntax errors reported by yacc. It is not yet used for errors (warnings) generated here. It would require the location to be passed to each function called from idl.y, in case there is not another method of sharing it through the context. One reason for to dispose this, is also that it might have to be changed again when preprocessing has been implemented, because we probably also would like to include the file name in the location. Either the lexer or the parser will need to process the %line directives produced by the preprocessor.

Reply:
I'm currently in the process of integrating the preprocessor into `idlc`. I don't think [location tracking](https://www.gnu.org/software/bison/manual/html_node/Tracking-Locations.html) will change with regards to anything in `tt_create.c`. Having said that, I do agree changes will eventually be required to `tt_create.*` for the compiler, yes. For the first iteration of `idlc`, I'll try to keep the changes to the "rest" to a minimum, but I really think you should have a look at the example I provided in `shift-reduce.zip` above. I may be overlooking something, but I think some things could be done in a more pragmatic way, maybe eliminating the need to pass locations to any function outside `idl.y` all together. I wouldn't focus on that too much for now though, it's probably a lot easier to have a working tool first to see how things like this can be implemented best.

Reply:
@FransFaase, the whole process is taking a little longer than I anticipated, but I've created a fairly complete demonstration of how to use locations (including filenames) [here](https://github.com/k0ekk0ek/lloc). I've also included basics for the `#pragma` directive. I'm hoping to get locations in this week and then move on to doing the rest of the integration of the preprocessor. Anyway, maybe the example will give you an idea of how it's going to work/look.